### ISSUE TITLE
pkgcheck scan: incorporate repo path (or path hash) in cache location

### ISSUE THREAD

**Issue by ezzieyguywuf** [pkgcore/pkgcheck#321]
I think the issue must be due to the `origin` remote tracking against the `dev` branch rather then `master`. I bet there's probably something hard coded in the pkgcheck code.

Here's what happens when I try to scan:

```sh
pkgcheck scan: error: failed running git log: fatal: Invalid revision range e5071aceb25da10a163b0216d140f4b93d56bd64..origin/HEAD
```

**Comment by ajakk** [comment]
I'm not sure this issue is specific to GURU, I get something very similar on the Gentoo repository:

```
gentoo $ pkgcheck scan
pkgcheck scan: error: failed running git log: fatal: Invalid revision range d742152a1d06293a6655ddc7402ee11869ab8c64..origin/HEAD
```

This commit hash isn't in gentoo.git, and over time the hash can change for some reason. This issue has persisted for a month or two now; it was a different hash when it started. Wolfgang's hash also isn't in the GURU repository. These hashes do come from somewhere though, the sync repository:

https://gitweb.gentoo.org/repo/sync/gentoo.git/commit/?id=e5071aceb25da10a163b0216d140f4b93d56bd64
https://gitweb.gentoo.org/repo/sync/gentoo.git/commit/?id=d742152a1d06293a6655ddc7402ee11869ab8c64

I have no clue why they'd be hard-coded anywhere, but:

```
~ $ qlist pkgcheck snakeoil pkgcore | xargs grep -E "(e5071aceb25da10a163b0216d140f4b93d56bd64|d742152a1d06293a6655ddc7402ee11869ab8c64)"
~ $
```

Bizzarely, I *am* able to `pkgcheck scan` the GURU repository, but I'm *not* able to scan a fresh clone of gentoo.git. In testing this I notice the hash changed again, so maybe this is related to cache updates between scanning the GURU and Gentoo repositories? The hash is equal to HEAD in my local clone of the sync repository in `/var/db/repos/gentoo`, dec0e031027a334fa8e168fa6656629eceebe02b.

Happy to see I'm not the only one hitting something like this! Been baffling me for a while now.

**Comment by radhermit** [comment]
> I think the issue must be due to the `origin` remote tracking against the `dev` branch rather then `master`. I bet there's probably something hard coded in the pkgcheck code.

There is no hardcoded branch usage in pkgcheck, only the refs ``HEAD`` and ``origin/HEAD`` which generally should exist if the checkout is being actively committed to.

**Comment by radhermit** [comment]
The first thing I'd try would be to totally regenerate a clean git cache for the targeted repo with something similar to ``pkgcheck cache -fu -t git -r gentoo``. Alternatively, all caches can be blown away with ``pkgcheck cache -R`` (or more selectively using -t/-r options) after which they should be regenerated automatically on the next run.

When the git cache is built it stores the current hash for the ``origin/HEAD`` ref and uses that to create the ref used for future updates, i.e. ``prev_hash..origin/HEAD``. This entirely assumes the relevant git history isn't being mangled in some fashion between cache updates.

**Comment by ajakk** [comment]
Yes, killing the cache (`pkgcheck cache -R`) and scanning again ('pkgcheck scan`) works. `pkgcheck scan --cache=-git` also worked.

Why would it cache a commit outside of the repository it's operating on, though?

**Comment by radhermit** [comment]
> Why would it cache a commit outside of the repository it's operating on, though?

As I said previously, either the history was mangled at some point or there is/was a bug in git cache creation. In all likelihood, it's some combination of the two, i.e. the current git cache support doesn't work well alongside ``git checkout`` usage to bounce between historical commits, changing the origin remote, or other actions between scan runs.

**Comment by radhermit** [comment]
Closing because I can't reproduce this using a git repo with linear history handling. Feel free to reopen with more information including some simple reproduction steps via git and pkgcheck commands to run post repo clone.

**Comment by ezzieyguywuf** [comment]
I can confirm - the following sequence for me will duplicate the error:

```sh
cd path/to/gentoo/repo
pkgcheck cache -R # start from scratch
pkgcheck scan # this rebuilds the cache - works fine
cd path/to/guru/repo
pkgcheck scan # crash - pkgcheck scan: error: failed running git log: fatal: Invalid revision range ....
pkgcheck cache -R # maybe it was just an old cache or something
pkgcheck scan # nice, it's fixed
cd path/to/gentoo/repo # let's make sure this is still good....
pkgcheck scan # nope! pkgcheck scan: error: failed running git log: fatal: Invalid revision range 
```

I think maybe the two caches are corrupting each other -probably they should be in two different directories but for some reason are being stored int he same place?

**Comment by AndrewAmmerlaan** [comment]
This happens for me as well, every time when switching between repos (not just ::guru).

What appears to be happening is that when you're working on e.g. ::guru, pkgcheck caches the sync-friendly version of ::gentoo set in repos.conf (since ::guru has `masters=gentoo`). When you're working on ::gentoo pkgcheck generates the cache directly from that repo (i.a.w. the dev version of ::gentoo), this version of the repo is naturally always going to be ahead of the sync-friendly repo that has the pre-generated md5-cache. Therefore, if you now switch back to working on e.g. ::guru, the cache is going to reference commits that do not yet exist in the sync-friendly version that pkgcheck wants to use. Hence the invalid revision range error, the cache was made for a revision range that is newer (i.a.w. the range from the dev version of::gentoo) then what is available in the repository that is being used (the sync-friendly version of ::gentoo).

**Comment by radhermit** [comment]
Reopening and renaming issue.

**Comment by radhermit** [comment]
Probably the easiest thing would be to store the path of the repo in the cache object and compare before rebuild/update, skipping repos that don't match since git cache support shouldn't be needed for repos underlying overlays.

**Comment by ajakk** [comment]
This is still reproducible when two repos with the same name are, in fact, different repos. For example with https://github.com/gentoo/gentoo and https://github.com/gentoo-mirror/gentoo.

**Comment by thesamesam** [comment]
See also https://github.com/pkgcore/pkgcheck/pull/528 (https://github.com/pkgcore/pkgcheck/commit/0e22d6f56b735b5b7449e5e3f800685bb54651d4).
